apps/emqx_rule_engine/src/emqx_rule_maps.erl

```diff
@@ -129,6 +129,15 @@ general_find({index, _}, List, _OrgData, Handler) when not is_list(List) ->
 
 do_put({key, Key}, Val, Map, _OrgData) when is_map(Map) ->
     maps:put(Key, Val, Map);
+do_put({key, Key}, Val, Data, _OrgData) when is_binary(Data) ->
+    case emqx_utils_json:safe_decode(Data, [return_maps]) of
```

Comment:
Just a thought. There seems to be more and more places where we pay for such flexibility with a few wasted CPU cycles here (since this clause is likely to be hit). Given this is part of a hot path, it might be costly. I'd honestly try to transfer this responsibility out of this module to those modules that know for sure they're dealing with JSON (as a separate effort obviously).

Reply:
Indeed, perhaps we could attempt the decoding once when the payload enters the rule engine to avoid this nested decoding, which can lead to the issue you commented below of inadvertently decoding what should be a string.